File: src/tools_savings_logger.py
#!/usr/bin/env python3
"""Lightweight tool-call logger for non-server tool paths (file router, file_tools_mcp.js).

Inserts a row into the `tool_calls` table inside the same SQLite database the
konoha MCP server writes to. The DB lives at ~/.konoha/skills.db.

Usage:
    python3 tools_savings_logger.py <tool_name> <query_json> <returned_bytes>
"""
import sys
import os
import sqlite3
import glob

HOME = os.path.expanduser("~")
DB_PATH = os.path.join(HOME, ".konoha", "skills.db")

DEFAULT_BASELINE = 550000  # ~140k tokens — matches server.py fallback

# Path constants matching server.py for client detection
ANTIGRAVITY_CLI_BRAIN = os.path.join(HOME, ".gemini", "antigravity-cli", "brain")
ANTIGRAVITY_IDE_BRAIN = os.path.join(HOME, ".gemini", "antigravity-ide", "brain")
CURSOR_PROJECTS = os.path.join(HOME, ".cursor", "projects")
CLAUDE_PROJECTS = os.path.join(HOME, ".claude", "projects")


def detect_active_client():
    """Detect which MCP client invoked this tool (mimics server.py logic)."""
    try:
        # Check environment variable first to distinguish CLI (agy) vs IDE (antigravity)
        conv_id = os.environ.get("ANTIGRAVITY_CONVERSATION_ID")
        if conv_id:
            cli_dir = os.path.join(ANTIGRAVITY_CLI_BRAIN, conv_id)
            if os.path.isdir(cli_dir):
                return "agy"
            ide_dir = os.path.join(ANTIGRAVITY_IDE_BRAIN, conv_id)
            if os.path.isdir(ide_dir):
                return "antigravity"

        if os.environ.get("OPENCODE_CLIENT") == "1" or os.environ.get("OPENCODE_SESSION") == "1":
            return "opencode"

        if os.environ.get("COMMANDCODE_CLIENT") == "1" or os.environ.get("COMMANDCODE_SESSION") == "1":
            return "commandcode"

        if os.environ.get("CLAUDE_CODE_CHILD_SESSION") == "1":
            return "claudecode"

        if conv_id:
            return "antigravity"

        brain_dirs = [
            ANTIGRAVITY_IDE_BRAIN,
            ANTIGRAVITY_CLI_BRAIN,
            CURSOR_PROJECTS,
            CLAUDE_PROJECTS,
        ]
        COMMANDCODE_PROJECTS = os.path.join(HOME, ".commandcode", "projects")
        if os.path.isdir(COMMANDCODE_PROJECTS):
            brain_dirs.append(COMMANDCODE_PROJECTS)

        all_files = []
        for brain_dir in brain_dirs:
            if not os.path.isdir(brain_dir):
                continue
            if "cursor" in brain_dir:
                pattern_transcript = os.path.join(brain_dir, "*", "agent-transcripts", "*", "*.jsonl")
                all_files.extend(glob.glob(pattern_transcript))
            elif "claude" in brain_dir:
                pattern_transcript = os.path.join(brain_dir, "*", "*.jsonl")
                all_files.extend(glob.glob(pattern_transcript))
            else:
                pattern_prompt = os.path.join(brain_dir, "*", "prompt.md")
                pattern_transcript = os.path.join(brain_dir, "*", ".system_generated", "logs", "transcript.jsonl")
                all_files.extend(glob.glob(pattern_prompt) + glob.glob(pattern_transcript))

        if not all_files:
            return "antigravity"

        all_files.sort(key=lambda x: os.path.getmtime(x), reverse=True)
        most_recent = all_files[0]

        if "cursor" in most_recent:
            return "cursor"
        elif "commandcode" in most_recent.lower():
            return "commandcode"
        elif "claudecode" in most_recent.lower() or "claude" in most_recent.lower():
            return "claudecode"
        elif "antigravity-cli" in most_recent:
            return "agy"
        else:
            return "antigravity"
    except Exception:
        pass
    return "antigravity"


def log(tool: str, query: str, returned_bytes: int, client: str = None) -> None:
    """Insert a single tool_calls row.

    baseline = sum of all skill byte sizes (proxy for "library would have
    been loaded"). Saving = baseline - returned_bytes. Always writes the
    actual returned_bytes so the dashboard shows real numbers, not zeros.
    """
    if client is None:
        client = detect_active_client()

    conn = sqlite3.connect(DB_PATH)
    cur = conn.cursor()
    try:
        baseline = DEFAULT_BASELINE
        try:
            row = cur.execute("SELECT SUM(byte_size) FROM skills").fetchone()
            if row and row[0]:
                baseline = int(row[0])
        except Exception:
            pass

        bytes_saved = max(baseline - returned_bytes, 0)
        tokens_saved = bytes_saved // 4

        cur.execute(
            """INSERT INTO tool_calls
               (tool, query, returned_bytes, total_library_bytes,
                bytes_saved, tokens_saved, client)
               VALUES (?, ?, ?, ?, ?, ?, ?)""",
            (tool[:200], query[:2000], returned_bytes,
             baseline, bytes_saved, tokens_saved, client),
        )
        conn.commit()
    finally:
        conn.close()


if __name__ == "__main__":
    if len(sys.argv) < 4:
        sys.stderr.write(
            "usage: tools_savings_logger.py <tool> <query> <returned_bytes> [client]\n"
        )
        sys.exit(1)
    try:
        client = sys.argv[4] if len(sys.argv) > 4 else None
        log(sys.argv[1], sys.argv[2], int(sys.argv[3]), client)
    except Exception as exc:
        # Never crash the caller — the router wraps us in try/ignore.
        sys.stderr.write(f"[tools_savings_logger] {exc}\n")
        sys.exit(0)

File: src/tools_savings_logger.js
#!/usr/bin/env node
/**
 * Lightweight tool-call logger for non-server tool paths (file router, file_tools_mcp.js).
 * Pure Node.js replacement for src/tools_savings_logger.py.
 * Inserts a row into the `tool_calls` table inside ~/.konoha/skills.db.
 */
const fs = require('fs');
const path = require('path');
const os = require('os');
const { getDb, DB_PATH } = require('./db');

const HOME = os.homedir();
const DEFAULT_BASELINE = 550000; // ~140k tokens — matches server.py/server.js fallback
const ANTIGRAVITY_CLI_BRAIN = path.join(HOME, '.gemini', 'antigravity-cli', 'brain');
const ANTIGRAVITY_IDE_BRAIN = path.join(HOME, '.gemini', 'antigravity-ide', 'brain');

function detectActiveClient() {
  try {
    const activeOverride = (process.env.ACTIVE_CLIENT || process.env.KONOHA_CLIENT || '').toLowerCase().trim();
    if (activeOverride) {
      if (activeOverride.includes('codex') || activeOverride.includes('openai')) return 'codex';
      if (activeOverride.includes('commandcode') || activeOverride.includes('command-code')) return 'commandcode';
      if (activeOverride.includes('opencode')) return 'opencode';
      if (activeOverride.includes('claude')) return 'claudecode';
      if (activeOverride.includes('cursor')) return 'cursor';
      if (activeOverride === 'pi') return 'pi';
      if (activeOverride.includes('agy') || activeOverride.includes('antigravity-cli')) return 'agy';
      if (activeOverride.includes('antigravity') || activeOverride.includes('ide')) return 'antigravity';
    }

    const convId = process.env.ANTIGRAVITY_CONVERSATION_ID;
    if (convId) {
      if ((process.env.ANTIGRAVITY_LS_VERSION || '').startsWith('cli') || (process.env.ANTIGRAVITY_AGENTAPI_EXE || '').includes('agy')) return 'agy';
      const cliDir = path.join(ANTIGRAVITY_CLI_BRAIN, convId);
      if (fs.existsSync(cliDir) && fs.statSync(cliDir).isDirectory()) return 'agy';
      const ideDir = path.join(ANTIGRAVITY_IDE_BRAIN, convId);
      if (fs.existsSync(ideDir) && fs.statSync(ideDir).isDirectory()) return 'antigravity';
      return 'agy';
    }

    if (process.env.OPENCODE_CLIENT === '1' || process.env.OPENCODE_SESSION === '1') return 'opencode';
    if (process.env.COMMANDCODE_CLIENT === '1' || process.env.COMMANDCODE_SESSION === '1') return 'commandcode';
    if (process.env.CLAUDE_CODE_CHILD_SESSION === '1') return 'claudecode';
    if (process.env.PI_CODING_AGENT === 'true' || process.env.PI_SESSION_FILE || process.env.PI_SESSION_ID) return 'pi';

    // No hard session signal: attribute HONESTLY instead of guessing.
    // The previous filesystem heuristic (glob every client's session dirs,
    // pick the most-recently-modified file) misattributed thousands of calls
    // made by CLI commands, test harnesses and scripts to whatever client
    // happened to have a fresh session file — corrupting the per-provider
    // breakdown in `konoha savings`. Provider rows must only contain calls
    // from sessions with a verified client signal.
    return 'unattributed';
  } catch (_) {
    return 'unattributed';
  }
}

function log(tool, query, returnedBytes, client = null, baselineBytes = null) {
  if (!client) {
    client = detectActiveClient();
  }
  const conn = getDb();
  try {
    let baseline = 0;
    if (baselineBytes !== null && baselineBytes !== undefined && Number(baselineBytes) > 0) {
      baseline = Number(baselineBytes);
    // aislop-ignore-next-line ai-slop/hardcoded-id (tool/provider NAME list, not a deployment identifier)
    } else if (tool === 'token_efficient_grep') {
      baseline = Math.max(Number(returnedBytes) || 0, 150000);
    } else if (tool === 'find_files_clean') {
      baseline = Math.max(Number(returnedBytes) || 0, 250000);
    } else if (['read_file_head', 'read_file_range', 'file_info', 'get_file_structure'].includes(tool)) {
      let resolvedSize = 0;
      try {
        let rawTarget = null;
        if (typeof query === 'string') {
          try {
            const parsed = JSON.parse(query);
            rawTarget = parsed.path || parsed.file_path || parsed.filepath || parsed.dir;
          } catch (_) {
            rawTarget = query;
          }
        }
        if (rawTarget) {
          const checkPath = path.isAbsolute(rawTarget) ? rawTarget : path.resolve(process.cwd(), rawTarget);
          if (fs.existsSync(checkPath)) {
            const st = fs.statSync(checkPath);
            resolvedSize = st.isFile() ? st.size : 100000;
          }
        }
      } catch (_) {
        // Ignore file stat errors
      }
      baseline = Math.max(Number(returnedBytes) || 0, resolvedSize);
    } else if (['find_skill', 'find_skills'].includes(tool)) {
      baseline = Math.max(Number(returnedBytes) || 0, 15000);
      try {
        const words = String(query || '').replace(/[^\w\s]/g, ' ').trim().split(/\s+/).filter(Boolean);
        if (words.length > 0) {
          const lk = '%' + words.join('%') + '%';
          const sumRow = conn.prepare(`
            SELECT COALESCE(SUM(byte_size), 0) as total_matched
            FROM (SELECT byte_size FROM skills WHERE tags LIKE ? OR name LIKE ? OR skill_name LIKE ? LIMIT 5)
          `).get(lk, lk, lk);
          if (sumRow && sumRow.total_matched > 0) {
            baseline = Math.max(Number(returnedBytes) || 0, Number(sumRow.total_matched));
          }
        }
      } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }
    } else if (tool === 'get_skill') {
      baseline = Number(returnedBytes) || 0;
      try {
        let skillTarget = query;
        if (typeof query === 'string' && query.startsWith('{')) {
          try { skillTarget = JSON.parse(query).name || query; } catch (_) { /* best-effort json parse fallback */ }
        }
        const sRow = conn.prepare('SELECT byte_size FROM skills WHERE name = ? OR name LIKE ? LIMIT 1').get(skillTarget, `%/${skillTarget}`);
        if (sRow && sRow.byte_size) {
          baseline = Math.max(Number(returnedBytes) || 0, Number(sRow.byte_size));
        }
      } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }
    } else if (['list_skills', 'optimize_report'].includes(tool)) {
      // Baseline rationale: Unpruned listing across ~138 skills requires loading raw frontmatter metadata
      // (averaging ~250 bytes each = ~34.5 KB, standardized to 35,000 bytes baseline). list_skills parses
      // and delivers concise JSON summaries (~15.7 KB), yielding ~55% empirical token reduction.
      baseline = Math.max(Number(returnedBytes) || 0, 35000);
    } else {
      baseline = Number(returnedBytes) || 0;
    }

    const retBytes = Number(returnedBytes) || 0;
    const bytesSaved = Math.max(baseline - retBytes, 0);
    const tokensSaved = Math.floor(bytesSaved / 4);

    conn.prepare(`
      INSERT INTO tool_calls
      (tool, query, returned_bytes, total_library_bytes, bytes_saved, tokens_saved, client)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `).run(
      String(tool || '').slice(0, 200),
      String(query || '').slice(0, 2000),
      retBytes,
      baseline,
      bytesSaved,
      tokensSaved,
      client
    );
  } catch (_) {
    // Fail silently
  } finally {
    try { conn.close(); } catch (_) { /* intentional best-effort fallback: failure here must never crash the CLI/MCP runtime */ }
  }
}

if (require.main === module) {
  if (process.argv.length < 5) {
    process.stderr.write('usage: tools_savings_logger.js <tool> <query> <returned_bytes> [client] [baseline_bytes]\n');
    process.exit(1);
  }
  try {
    const tool = process.argv[2];
    const query = process.argv[3];
    const returnedBytes = parseInt(process.argv[4], 10) || 0;
    const client = process.argv[5] || null;
    const baselineBytes = process.argv[6] ? parseInt(process.argv[6], 10) : null;
    log(tool, query, returnedBytes, client, baselineBytes);
  } catch (exc) {
    process.stderr.write(`[tools_savings_logger] ${exc.message}\n`);
    process.exit(0);
  }
}

module.exports = {
  log,
  detectActiveClient,
  DEFAULT_BASELINE,
  DB_PATH
};
